### PR TITLE
Enforce mutual exclusivity among view, materialized view, and schema in BigQuery table config

### PR DESCRIPTION
**Pull request by wj-chen**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/14220.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
bigquery: fixed view and materialized view creation when schema is specified
```

### CODE REVIEW

**Review by melinath — CHANGES_REQUESTED** [review]
Sorry for the delayed review - lost track of this. I'm concerned about this implementation because it seems like it is intentionally not updating the schema field.

After a user runs `terraform apply`, the API resource should match what is in the Terraform configuration - so the schema field also needs to be set properly _during create_ for Terraform, even though the API only supports setting it via update. From the ticket, it sounds like the problem was that the update may not have worked because of a problem on the API side? Please let me know if you think I'm missing something :-)

**Comment by wj-chen** [comment]
> 

Oops I requested a re-review before seeing your comment. Thanks for taking a look!

In short, "view"/"materialized_view" and "schema" should be mutually exclusive, hence the (unchanged in this PR) behavior of setting "schema" to nil for the create request. If "schema" is specified when "view"/"materialized_view" is present, it should be ignored, hence the change in this PR (there is a buganizer issue describing why that needs to be, can I link it here?). So in this case, by nature, the resource to be created and what's shown in the Terraform config will differ.

If it's more in line with the style standard - instead of ignoring, we can outright fail the `terraform apply` if the conflicting fields are set. If you could point me to an example of how that's implemented and tested that would help. THis is my first Terraform provider change.

**Review by melinath — CHANGES_REQUESTED** [review]
more context at b/280326829 (yes, fine to share - happy to move discussion there if needed)

If the fields are mutually exclusive (not just needing an additional update call), you can add [`ConflictsWith`](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-methods) to the schema to make sure that users are alerted as early as possible about the conflict. It will show up as a validation error at plan time.

That would be preferred over silently not sending the schema and then having a diff, since the configuration is essentially asking for something invalid.

Adding ConflictsWith would generally be considered a breaking change, and would therefore need to be held for the next major release, because it could cause a configuration that previously applied successfully to start failing.

However, if an invalid configuration would always cause the API to return an error (so that the configuration could never be successfully applied) or if it would always cause a permadiff (especially a destructive permadiff) then we may be able to make an exception and include the change in a minor release.

I'll need to investigate this more closely myself before merge, but I wanted to get your input on whether you think it would be better to wait for a major release vs. trying to move forward now based on what I said above.

**Comment by wj-chen** [comment]
> more context at b/280326829 (yes, fine to share - happy to move discussion there if needed)
> 
> If the fields are mutually exclusive (not just needing an additional update call), you can add [`ConflictsWith`](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-methods) to the schema to make sure that users are alerted as early as possible about the conflict. It will show up as a validation error at plan time.
> 
> That would be preferred over silently not sending the schema and then having a diff, since the configuration is essentially asking for something invalid.

Thank you for the pointer. ConflictsWith looks very promising and I'll look into it. As mentioned in the bug, throwing a validation error in this case is the current behavior of other clients e.g. CLI.

> 
> Adding ConflictsWith would generally be considered a breaking change, and would therefore need to be held for the next major release, because it could cause a configuration that previously applied successfully to start failing.
> 
> However, if an invalid configuration would always cause the API to return an error (so that the configuration could never be successfully applied) or if it would always cause a permadiff (especially a destructive permadiff) then we may be able to make an exception and include the change in a minor release.
> 
> I'll need to investigate this more closely myself before merge, but I wanted to get your input on whether you think it would be better to wait for a major release vs. trying to move forward now based on what I said above.

I think this can wait until the next major release. The extraneous update call may or may not fail depending on the exact schema, so it's not an always-return-an-error situation, but I'd like to understand what you menat by "permadiff"?

Also general question about release cadence - for a newly submitted change, how long do we expect it to be released to terraform-google-providers and terraform-google-providers-beta as part of a minor or major release?

**Comment by melinath** [comment]
regarding the questions - I'll reach out internally.

**Comment by wj-chen** [comment]
Update: I was able to verify locally the desired behavior using `ConflictWith`, now onto adding tests for it but haven't found good examples that test pre-apply validations. Do you have any pointers?

**Comment by melinath** [comment]
Generally we don't add tests to make sure that ConflictsWith is honored by Terraform. But you could probably test it with [ExpectError](https://developer.hashicorp.com/terraform/plugin/testing/testing-patterns#expecting-errors-or-non-empty-plans) if you want to. This would be best as part of an existing test if bigquery resources take a while to set up.

**Comment by wj-chen** [comment]
> Generally we don't add tests to make sure that ConflictsWith is honored by Terraform. But you could probably test it with [ExpectError](https://developer.hashicorp.com/terraform/plugin/testing/testing-patterns#expecting-errors-or-non-empty-plans) if you want to. This would be best as part of an existing test if bigquery resources take a while to set up.

Thanks again! I updated this PR and included new tests. Since the new tests are expected to fail in the schema validation stage, they shouldn't take a while to set up since they don't actually create any resource.

**Review by melinath — APPROVED** [review]
LGTM - this will be much more clear! I've added the referenced ticket to the 5.0.0 milestone. Please reach out at the end of July to get this merged.

**Review by melinath — CHANGES_REQUESTED** [review]
Please resolve the conflicts in mmv1/third_party/terraform/resources/resource_bigquery_table.go - you could do this by rebasing onto the major release branch or be merging the major release branch into this branch.

Note - I've [changed the base branch of this PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request). Since the PR is opened from your main branch, you'll need to make all updates for this PR to your main branch. In the future I'd recommend creating new branches as described in [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow) so that your main branch (or any feature release branches) can track the related branch in this repository.

**Comment by wj-chen** [comment]
I need to clean up my fork of the repo. Will ping here again when it's ready.

**Comment by wj-chen** [comment]
The PR has been cleaned up. VCR is failing for a new test:
```
=== CONT  TestAccBigQueryTable_MaterializedView_WithView
    test_utils.go:98: no cassette found on disk for TestAccBigQueryTable_MaterializedView_WithView, please replay this testcase in recording mode - open /go/src/github.com/hashicorp/terraform-provider-google-beta/fixtures/TestAccBigQueryTable_MaterializedView_WithView.seed: no such file or directory
--- FAIL: TestAccBigQueryTable_MaterializedView_WithView (0.00s)
```
I saw this in another PR of mine (https://github.com/GoogleCloudPlatform/magic-modules/pull/8460) and was told to skip VCR. Should I do the same here?

**Comment by melinath** [comment]
hmm, I'm not sure - asking a teammate

**Comment by melinath** [comment]
I don't see a reason why this would not be possible to record - it's testing that the resource fails to create. There's nothing that looks nondeterministic about the tests either - I'm going to rerun the tests in case that helps.

**Review by melinath — CHANGES_REQUESTED** [review]
> I saw this in another PR of mine (https://github.com/GoogleCloudPlatform/magic-modules/pull/8460) and was told to skip VCR. Should I do the same here?

@wj-chen okay, so, for https://github.com/GoogleCloudPlatform/magic-modules/pull/8460 - generally we add unit tests for validation functions rather than acceptance tests; having an acceptance test that confirms the validation function is actually in use is also fine, but it will be harder to cover all the potential edge cases. (FWIW I'd recommend adding some unit tests for validateBigQueryTableSchema.)

However, for this PR, there's no function to test, so acceptance tests is the only option - and we don't currently support running acceptance tests that don't make API requests on PRs using VCR. **So, skipping during VCR runs would be the correct step for now.**

(I've opened https://github.com/hashicorp/terraform-provider-google/issues/15427 to track supporting this use case better. We'll still run the tests in our nightly builds.)

File: mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go
Comment (melinath) [inline mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go:507]: It looks like this is the wrong schema field - this is `external_data_configuration.schema`

File: mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go
Comment (melinath) [inline mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go:776]: I think this is the schema field that should be mutually exclusive?

**Review by melinath — CHANGES_REQUESTED** [review]
In addition to skipping the tests in VCR - it looks like the wrong schema field is marked as conflicting.

File: mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go
Comment (wj-chen) [reply mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go]: Thanks for the catch, fixed

File: mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go
Comment (wj-chen) [reply mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go:775]: Thanks for the catch, fixed

**Comment by wj-chen** [comment]
> It would also be great if you could prep a pr to add this change to https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/guides/version_5_upgrade.html.markdown on the main branch of magic-modules

Is this like a changelog? I'll do so in a separate PR.

**Comment by melinath** [comment]
> Is this like a changelog? I'll do so in a separate PR.

yeah it'll need to be a separate PR since it'll be a change to the main branch rather than the major release branch :-) It's the guide that users will follow to upgrade their configurations leading up to 5.0.0. For comparison, here's the 4.0.0 upgrade guide: https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/version_4_upgrade

(Rerunning tests /gcbrun)

**Comment by wj-chen** [comment]
> @wj-chen Can you also add an upgrade guide entry for the breaking change(s) in #8460 ? Feel free to ping/tag me in the PR so I can review.

For sure, will do.

**Comment by wj-chen** [comment]
> > Is this like a changelog? I'll do so in a separate PR.
> 
> yeah it'll need to be a separate PR since it'll be a change to the main branch rather than the major release branch :-) It's the guide that users will follow to upgrade their configurations leading up to 5.0.0. For comparison, here's the 4.0.0 upgrade guide: https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/version_4_upgrade
> 
> (Rerunning tests /gcbrun)

Thanks for the resources! Will do once this is merged.

**Comment by melinath** [comment]
Manual test run since these can't run in VCR: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloud_ProviderGoogleCloudMmUpstream/426090?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

**Review by melinath — APPROVED** [review]
test passed in CI, lgtm

**Comment by melinath** [comment]
Please be sure to add this to the [version 5 upgrade guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/guides/version_5_upgrade.html.markdown)! Thanks!

**Comment by wj-chen** [comment]
> Please be sure to add this to the [version 5 upgrade guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/guides/version_5_upgrade.html.markdown)! Thanks!

Thank you very much! Will get on that next.

**Comment by wj-chen** [comment]
> > Please be sure to add this to the [version 5 upgrade guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/guides/version_5_upgrade.html.markdown)! Thanks!
> 
> Thank you very much! Will get on that next.

Upgrade doc update has been created at #8607.

**Comment by wj-chen** [comment]
> #8460

@c2thorn PR for adding upgrade notes for #8460: #8611.

**Pull request merged**
